src/v/bytes/oncore.h

```diff
@@ -26,16 +28,29 @@ class oncore final {
     // owner shard set on construction
     oncore();
 
-    void verify_shard_source_location(const char* file, int linenum) const;
+    void assert_shard_source_location(
+      const char* = get_file_basename(),
+      unsigned line = std::source_location::current().line()) const;
+    void log_shard_source_location(
+      const char* = get_file_basename(),
+      unsigned line = std::source_location::current().line()) const;
 
 private:
     shard_id_type _owner_shard;
 };
 
-// Next function should be replace with source_location in c++20 very soon
+// Debug builds assert, no checks in release builds
 // NOLINTNEXTLINE
 #define oncore_debug_verify(member)                                            \
     do {                                                                       \
-        expression_in_debug_mode((member).verify_shard_source_location(        \
-          get_file_basename(), __LINE__));                                     \
+        expression_in_debug_mode((member).assert_shard_source_location());     \
+    } while (0)
+
+// Debug builds assert, but release builds emit error logs
+// NOLINTNEXTLINE
+#define oncore_soft_verify(member)                                             \
```

Comment:
> but release builds emit error logs

Is there a particular use case for this? It almost seems backwards: an on-core violation is generally a logic error and will be associated with race conditions. It's more likely to lead to corruption in release builds where we don't have a sanitizer, for example.

Reply:
Would you rather just have a case where we unconditionally assert? I would prefer that.

Reply:
Changed to just always assert.

Reply:
>Would you rather just have a case where we unconditionally assert? I would prefer that.

I thought that is what `bytes/oncore.cc` was already doing?

Reply:
Well yes, but all usages wrap that in the debug_expression macro currently. So none of these checks actually happen in release builds

Reply:
>Well yes, but all usages wrap that in the debug_expression macro currently. So none of these checks actually happen in release builds

oh, right. yeh, i can't really think of any hot paths that are so hot to want to disable a critical check.